{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Data_from_Biomart_solutions.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyOFshtVq88L8U11OrIlM0Z4",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/araldi/HS21---Big-Data-Analysis-in-Biomedical-Research-376-1723-00L-/blob/main/Week11/Data_from_Biomart_solutions.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zrKEOAZHFzeE"
      },
      "source": [
        "Biomart provides info on genes, polymorphisms, sequences, orthology, genomic features and other genetic information (on Homo sapiens and other organisms)."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "background_save": true,
          "base_uri": "https://localhost:8080/"
        },
        "id": "338Ii0wdF92_",
        "outputId": "6f2180ea-0461-4514-ad96-88feb94b46f6"
      },
      "source": [
        "pip install pybiomart"
      ],
      "execution_count": null,
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting pybiomart\n",
            "  Downloading pybiomart-0.2.0-py3-none-any.whl (10 kB)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from pybiomart) (2.23.0)\n",
            "Collecting requests-cache\n",
            "  Downloading requests_cache-0.8.1-py3-none-any.whl (44 kB)\n",
            "\u001b[K     |████████████████████████████████| 44 kB 2.5 MB/s \n",
            "\u001b[?25hRequirement already satisfied: future in /usr/local/lib/python3.7/dist-packages (from pybiomart) (0.16.0)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from pybiomart) (1.1.5)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas->pybiomart) (2018.9)\n",
            "Requirement already satisfied: numpy>=1.15.4 in /usr/local/lib/python3.7/dist-packages (from pandas->pybiomart) (1.19.5)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas->pybiomart) (2.8.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->pandas->pybiomart) (1.15.0)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->pybiomart) (3.0.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->pybiomart) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->pybiomart) (1.24.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->pybiomart) (2021.10.8)\n",
            "Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1\n",
            "  Downloading urllib3-1.25.11-py2.py3-none-any.whl (127 kB)\n",
            "\u001b[K     |████████████████████████████████| 127 kB 45.4 MB/s \n",
            "\u001b[?25hCollecting url-normalize<2.0,>=1.4\n",
            "  Downloading url_normalize-1.4.3-py2.py3-none-any.whl (6.8 kB)\n",
            "Requirement already satisfied: attrs<22.0,>=21.2 in /usr/local/lib/python3.7/dist-packages (from requests-cache->pybiomart) (21.2.0)\n",
            "Collecting cattrs<2.0,>=1.8\n",
            "  Downloading cattrs-1.8.0-py3-none-any.whl (24 kB)\n",
            "Requirement already satisfied: appdirs<2.0.0,>=1.4.4 in /usr/local/lib/python3.7/dist-packages (from requests-cache->pybiomart) (1.4.4)\n",
            "Installing collected packages: urllib3, url-normalize, cattrs, requests-cache, pybiomart\n",
            "  Attempting uninstall: urllib3\n",
            "    Found existing installation: urllib3 1.24.3\n",
            "    Uninstalling urllib3-1.24.3:\n",
            "      Successfully uninstalled urllib3-1.24.3\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.\u001b[0m\n",
            "Successfully installed cattrs-1.8.0 pybiomart-0.2.0 requests-cache-0.8.1 url-normalize-1.4.3 urllib3-1.25.11\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AtGOsLhcFq6V"
      },
      "source": [
        "import pandas as pd\n",
        "from pybiomart import Server"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dOYoRfayGbYs"
      },
      "source": [
        "server = Server(host='www.ensembl.org')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 260
        },
        "id": "mKM-1UbiGnG6",
        "outputId": "3c8578b7-269e-476f-e960-5d5de41ca0d7"
      },
      "source": [
        "server.list_marts()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>name</th>\n",
              "      <th>display_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSEMBL_MART_ENSEMBL</td>\n",
              "      <td>Ensembl Genes 104</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSEMBL_MART_MOUSE</td>\n",
              "      <td>Mouse strains 104</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSEMBL_MART_SEQUENCE</td>\n",
              "      <td>Sequence</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSEMBL_MART_ONTOLOGY</td>\n",
              "      <td>Ontology</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSEMBL_MART_GENOMIC</td>\n",
              "      <td>Genomic features 104</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>ENSEMBL_MART_SNP</td>\n",
              "      <td>Ensembl Variation 104</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>ENSEMBL_MART_FUNCGEN</td>\n",
              "      <td>Ensembl Regulation 104</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    name            display_name\n",
              "0   ENSEMBL_MART_ENSEMBL       Ensembl Genes 104\n",
              "1     ENSEMBL_MART_MOUSE       Mouse strains 104\n",
              "2  ENSEMBL_MART_SEQUENCE                Sequence\n",
              "3  ENSEMBL_MART_ONTOLOGY                Ontology\n",
              "4   ENSEMBL_MART_GENOMIC    Genomic features 104\n",
              "5       ENSEMBL_MART_SNP   Ensembl Variation 104\n",
              "6   ENSEMBL_MART_FUNCGEN  Ensembl Regulation 104"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Ziu2OIoYJze8"
      },
      "source": [
        "# Get information on genes"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tyFJLGJYGzzs"
      },
      "source": [
        "mart = server['ENSEMBL_MART_ENSEMBL']\n",
        "mart.list_datasets()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 408
        },
        "id": "cOfU3zuIHFDz",
        "outputId": "d4afb707-7284-46e5-e09a-ca33e903e021"
      },
      "source": [
        "dataset = mart['hsapiens_gene_ensembl']\n",
        "\n",
        "dataset.list_attributes()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>name</th>\n",
              "      <th>display_name</th>\n",
              "      <th>description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ensembl_gene_id</td>\n",
              "      <td>Gene stable ID</td>\n",
              "      <td>Stable ID of the Gene</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ensembl_gene_id_version</td>\n",
              "      <td>Gene stable ID version</td>\n",
              "      <td>Versionned stable ID of the Gene</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ensembl_transcript_id</td>\n",
              "      <td>Transcript stable ID</td>\n",
              "      <td>Stable ID of the Transcript</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ensembl_transcript_id_version</td>\n",
              "      <td>Transcript stable ID version</td>\n",
              "      <td>Versionned stable ID of the Transcript</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ensembl_peptide_id</td>\n",
              "      <td>Protein stable ID</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3192</th>\n",
              "      <td>sequence_genomic_coding_end</td>\n",
              "      <td>Genomic coding end</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3193</th>\n",
              "      <td>sequence_constitutive</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3194</th>\n",
              "      <td>cds_length</td>\n",
              "      <td>CDS Length</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3195</th>\n",
              "      <td>cds_start</td>\n",
              "      <td>CDS start</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3196</th>\n",
              "      <td>cds_end</td>\n",
              "      <td>CDS end</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>3197 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                               name  ...                             description\n",
              "0                   ensembl_gene_id  ...                   Stable ID of the Gene\n",
              "1           ensembl_gene_id_version  ...        Versionned stable ID of the Gene\n",
              "2             ensembl_transcript_id  ...             Stable ID of the Transcript\n",
              "3     ensembl_transcript_id_version  ...  Versionned stable ID of the Transcript\n",
              "4                ensembl_peptide_id  ...                                        \n",
              "...                             ...  ...                                     ...\n",
              "3192    sequence_genomic_coding_end  ...                                        \n",
              "3193          sequence_constitutive  ...                                        \n",
              "3194                     cds_length  ...                                        \n",
              "3195                      cds_start  ...                                        \n",
              "3196                        cds_end  ...                                        \n",
              "\n",
              "[3197 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tPO-gXSrHWaz"
      },
      "source": [
        "## Exercise\n",
        "Create a dictionary of gene names to Ensembl Gene IDs"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "NYdP952dFmZI"
      },
      "source": [
        "genes_for_dict_38 = dataset.query(\n",
        "                                  attributes=['ensembl_gene_id',\n",
        "                                              'description',\n",
        "                                              'external_gene_name'])\n",
        "\n",
        "dict_gene = {}\n",
        "for i in genes_for_dict_38.itertuples():\n",
        "    dict_gene[i._3] = i._1"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "sPtAAlF8HncI",
        "outputId": "829be30c-7f45-46dd-bc1d-f7a3b9d7d61c"
      },
      "source": [
        "# Example\n",
        "dict_gene['APOB']"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'ENSG00000084674'"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HTv6n_hEH27L"
      },
      "source": [
        "## Exercise \n",
        "Create a list of all the genes on chromosome 5, with gene start, gene end, gene description"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 351
        },
        "id": "beKTrF9UIWM6",
        "outputId": "096da29c-5485-4b3f-884d-c95b6bebd225"
      },
      "source": [
        "dataset.list_filters().head(10)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>name</th>\n",
              "      <th>type</th>\n",
              "      <th>description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>link_so_mini_closure</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>link_go_closure</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>link_ensembl_transcript_stable_id</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>gene_id</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>transcript_id</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>link_ensembl_gene_id</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>chromosome_name</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>start</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>end</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>band_start</td>\n",
              "      <td>drop_down_basic_filter</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                name                    type description\n",
              "0               link_so_mini_closure                    list            \n",
              "1                    link_go_closure                    text            \n",
              "2  link_ensembl_transcript_stable_id                    text            \n",
              "3                            gene_id                    text            \n",
              "4                      transcript_id                    text            \n",
              "5               link_ensembl_gene_id                    text            \n",
              "6                    chromosome_name                    text            \n",
              "7                              start                    text            \n",
              "8                                end                    text            \n",
              "9                         band_start  drop_down_basic_filter            "
            ]
          },
          "metadata": {},
          "execution_count": 27
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "y3EdI-Pa76z0"
      },
      "source": [
        "def genes_in_chromosome_of_interest(chromosome_number):\n",
        "  chrom_of_interest = dataset.query(filters={'chromosome_name' : str(chromosome_number) },\n",
        "                                    attributes=['ensembl_gene_id',\n",
        "                                                'description',\n",
        "                                                'external_gene_name',\n",
        "                                                'chromosome_name',\n",
        "                                                'start_position',\n",
        "                                                'end_position'])\n",
        "  chrom_of_interest = chrom_of_interest[chrom_of_interest['Gene name'].notna()]\n",
        "  return chrom_of_interest\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 408
        },
        "id": "XXIF8MeRIPaP",
        "outputId": "9025880d-9512-4cdd-ccdd-c4f632dd0615"
      },
      "source": [
        "genes_in_chromosome_of_interest(5)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene stable ID</th>\n",
              "      <th>Gene description</th>\n",
              "      <th>Gene name</th>\n",
              "      <th>Chromosome/scaffold name</th>\n",
              "      <th>Gene start (bp)</th>\n",
              "      <th>Gene end (bp)</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000153404</td>\n",
              "      <td>pleckstrin homology and RhoGEF domain containi...</td>\n",
              "      <td>PLEKHG4B</td>\n",
              "      <td>5</td>\n",
              "      <td>92151</td>\n",
              "      <td>189972</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000199540</td>\n",
              "      <td>Y RNA [Source:RFAM;Acc:RF00019]</td>\n",
              "      <td>Y_RNA</td>\n",
              "      <td>5</td>\n",
              "      <td>144087</td>\n",
              "      <td>144197</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000185028</td>\n",
              "      <td>leucine rich repeat containing 14B [Source:HGN...</td>\n",
              "      <td>LRRC14B</td>\n",
              "      <td>5</td>\n",
              "      <td>191495</td>\n",
              "      <td>196334</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>ENSG00000164366</td>\n",
              "      <td>coiled-coil domain containing 127 [Source:HGNC...</td>\n",
              "      <td>CCDC127</td>\n",
              "      <td>5</td>\n",
              "      <td>196868</td>\n",
              "      <td>218153</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>ENSG00000073578</td>\n",
              "      <td>succinate dehydrogenase complex flavoprotein s...</td>\n",
              "      <td>SDHA</td>\n",
              "      <td>5</td>\n",
              "      <td>218241</td>\n",
              "      <td>257082</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2981</th>\n",
              "      <td>ENSG00000274553</td>\n",
              "      <td>capicua transcriptional repressor pseudogene 1...</td>\n",
              "      <td>CICP15</td>\n",
              "      <td>5</td>\n",
              "      <td>181329049</td>\n",
              "      <td>181331839</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2982</th>\n",
              "      <td>ENSG00000250765</td>\n",
              "      <td>septin 14 pseudogene 5 [Source:HGNC Symbol;Acc...</td>\n",
              "      <td>SEPTIN14P5</td>\n",
              "      <td>5</td>\n",
              "      <td>181329241</td>\n",
              "      <td>181342213</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2983</th>\n",
              "      <td>ENSG00000230178</td>\n",
              "      <td>olfactory receptor family 4 subfamily F member...</td>\n",
              "      <td>OR4F3</td>\n",
              "      <td>5</td>\n",
              "      <td>181367268</td>\n",
              "      <td>181368262</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2984</th>\n",
              "      <td>ENSG00000251035</td>\n",
              "      <td>WBP1L pseudogene 4 [Source:HGNC Symbol;Acc:HGN...</td>\n",
              "      <td>WBP1LP4</td>\n",
              "      <td>5</td>\n",
              "      <td>181378695</td>\n",
              "      <td>181379038</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2985</th>\n",
              "      <td>ENSG00000213285</td>\n",
              "      <td>ribosomal protein L23a pseudogene 45 [Source:H...</td>\n",
              "      <td>RPL23AP45</td>\n",
              "      <td>5</td>\n",
              "      <td>181471963</td>\n",
              "      <td>181472430</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1837 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "       Gene stable ID  ... Gene end (bp)\n",
              "1     ENSG00000153404  ...        189972\n",
              "2     ENSG00000199540  ...        144197\n",
              "3     ENSG00000185028  ...        196334\n",
              "5     ENSG00000164366  ...        218153\n",
              "7     ENSG00000073578  ...        257082\n",
              "...               ...  ...           ...\n",
              "2981  ENSG00000274553  ...     181331839\n",
              "2982  ENSG00000250765  ...     181342213\n",
              "2983  ENSG00000230178  ...     181368262\n",
              "2984  ENSG00000251035  ...     181379038\n",
              "2985  ENSG00000213285  ...     181472430\n",
              "\n",
              "[1837 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-oE6IbV5J2dn"
      },
      "source": [
        "# Get information on SNPs"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "ftvvxYd_IUbc",
        "outputId": "e0fe30ca-26eb-4dc7-eff3-b884841336bc"
      },
      "source": [
        "mart = server['ENSEMBL_MART_SNP']\n",
        "mart.list_datasets()\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>name</th>\n",
              "      <th>display_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>mmulatta_structvar</td>\n",
              "      <td>Macaque Structural Variants (Mmul_10)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>fcatus_snp</td>\n",
              "      <td>Cat Short Variants (SNPs and indels excluding ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>hsapiens_structvar_som</td>\n",
              "      <td>Human Somatic Structural Variants (GRCh38.p13)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>oarambouillet_structvar</td>\n",
              "      <td>Sheep Structural Variants (Oar_rambouillet_v1.0)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>mdomestica_snp</td>\n",
              "      <td>Opossum Short Variants (SNPs and indels exclud...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>mmusculus_snp</td>\n",
              "      <td>Mouse Short Variants (SNPs and indels excludin...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>tguttata_snp</td>\n",
              "      <td>Zebra finch Short Variants (SNPs and indels ex...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>pabelii_snp</td>\n",
              "      <td>Orangutan Short Variants (SNPs and indels excl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>tnigroviridis_snp</td>\n",
              "      <td>Tetraodon Short Variants (SNPs and indels excl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>sscrofa_structvar</td>\n",
              "      <td>Pig Structural Variants (Sscrofa11.1)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>btaurus_structvar</td>\n",
              "      <td>Cow Structural Variants (ARS-UCD1.2)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>rnorvegicus_snp</td>\n",
              "      <td>Rat Short Variants (SNPs and indels excluding ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>hsapiens_structvar</td>\n",
              "      <td>Human Structural Variants (GRCh38.p13)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>oaries_structvar</td>\n",
              "      <td>Sheep (texel) Structural Variants (Oar_v3.1)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>ecaballus_snp</td>\n",
              "      <td>Horse Short Variants (SNPs and indels excludin...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>oaries_snp</td>\n",
              "      <td>Sheep (texel) Short Variants (SNPs and indels ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>sscrofa_snp</td>\n",
              "      <td>Pig Short Variants (SNPs and indels excluding ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>mmusculus_structvar</td>\n",
              "      <td>Mouse Structural Variants (GRCm39)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>mgallopavo_snp</td>\n",
              "      <td>Turkey Short Variants (SNPs and indels excludi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>hsapiens_snp</td>\n",
              "      <td>Human Short Variants (SNPs and indels excludin...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>drerio_structvar</td>\n",
              "      <td>Zebrafish Structural Variants (GRCz11)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>chircus_snp</td>\n",
              "      <td>Goat Short Variants (SNPs and indels excluding...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>btaurus_snp</td>\n",
              "      <td>Cow Short Variants (SNPs and indels excluding ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>drerio_snp</td>\n",
              "      <td>Zebrafish Short Variants (SNPs and indels excl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>clfamiliaris_structvar</td>\n",
              "      <td>Dog Structural Variants (CanFam3.1)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>scerevisiae_snp</td>\n",
              "      <td>Saccharomyces cerevisiae Short Variants (SNPs ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>ecaballus_structvar</td>\n",
              "      <td>Horse Structural Variants (EquCab3.0)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>mmulatta_snp</td>\n",
              "      <td>Macaque Short Variants (SNPs and indels exclud...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>clfamiliaris_snp</td>\n",
              "      <td>Dog Short Variants (SNPs and indels excluding ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>ggallus_snp</td>\n",
              "      <td>Chicken Short Variants (SNPs and indels exclud...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>oanatinus_snp</td>\n",
              "      <td>Platypus Short Variants (SNPs and indels exclu...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>nleucogenys_snp</td>\n",
              "      <td>Gibbon Short Variants (SNPs and indels excludi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>ptroglodytes_snp</td>\n",
              "      <td>Chimpanzee Short Variants (SNPs and indels exc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>oarambouillet_snp</td>\n",
              "      <td>Sheep Short Variants (SNPs and indels excludin...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>hsapiens_snp_som</td>\n",
              "      <td>Human Somatic Short Variants (SNPs and indels ...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                       name                                       display_name\n",
              "0        mmulatta_structvar              Macaque Structural Variants (Mmul_10)\n",
              "1                fcatus_snp  Cat Short Variants (SNPs and indels excluding ...\n",
              "2    hsapiens_structvar_som     Human Somatic Structural Variants (GRCh38.p13)\n",
              "3   oarambouillet_structvar   Sheep Structural Variants (Oar_rambouillet_v1.0)\n",
              "4            mdomestica_snp  Opossum Short Variants (SNPs and indels exclud...\n",
              "5             mmusculus_snp  Mouse Short Variants (SNPs and indels excludin...\n",
              "6              tguttata_snp  Zebra finch Short Variants (SNPs and indels ex...\n",
              "7               pabelii_snp  Orangutan Short Variants (SNPs and indels excl...\n",
              "8         tnigroviridis_snp  Tetraodon Short Variants (SNPs and indels excl...\n",
              "9         sscrofa_structvar              Pig Structural Variants (Sscrofa11.1)\n",
              "10        btaurus_structvar               Cow Structural Variants (ARS-UCD1.2)\n",
              "11          rnorvegicus_snp  Rat Short Variants (SNPs and indels excluding ...\n",
              "12       hsapiens_structvar             Human Structural Variants (GRCh38.p13)\n",
              "13         oaries_structvar       Sheep (texel) Structural Variants (Oar_v3.1)\n",
              "14            ecaballus_snp  Horse Short Variants (SNPs and indels excludin...\n",
              "15               oaries_snp  Sheep (texel) Short Variants (SNPs and indels ...\n",
              "16              sscrofa_snp  Pig Short Variants (SNPs and indels excluding ...\n",
              "17      mmusculus_structvar                 Mouse Structural Variants (GRCm39)\n",
              "18           mgallopavo_snp  Turkey Short Variants (SNPs and indels excludi...\n",
              "19             hsapiens_snp  Human Short Variants (SNPs and indels excludin...\n",
              "20         drerio_structvar             Zebrafish Structural Variants (GRCz11)\n",
              "21              chircus_snp  Goat Short Variants (SNPs and indels excluding...\n",
              "22              btaurus_snp  Cow Short Variants (SNPs and indels excluding ...\n",
              "23               drerio_snp  Zebrafish Short Variants (SNPs and indels excl...\n",
              "24   clfamiliaris_structvar                Dog Structural Variants (CanFam3.1)\n",
              "25          scerevisiae_snp  Saccharomyces cerevisiae Short Variants (SNPs ...\n",
              "26      ecaballus_structvar              Horse Structural Variants (EquCab3.0)\n",
              "27             mmulatta_snp  Macaque Short Variants (SNPs and indels exclud...\n",
              "28         clfamiliaris_snp  Dog Short Variants (SNPs and indels excluding ...\n",
              "29              ggallus_snp  Chicken Short Variants (SNPs and indels exclud...\n",
              "30            oanatinus_snp  Platypus Short Variants (SNPs and indels exclu...\n",
              "31          nleucogenys_snp  Gibbon Short Variants (SNPs and indels excludi...\n",
              "32         ptroglodytes_snp  Chimpanzee Short Variants (SNPs and indels exc...\n",
              "33        oarambouillet_snp  Sheep Short Variants (SNPs and indels excludin...\n",
              "34         hsapiens_snp_som  Human Somatic Short Variants (SNPs and indels ..."
            ]
          },
          "metadata": {},
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 653
        },
        "id": "tMyX9uH4KHDG",
        "outputId": "0346bfdf-5230-47a4-bf78-77c4772bb143"
      },
      "source": [
        "dataset = mart['hsapiens_snp']\n",
        "\n",
        "dataset.list_attributes().head(20)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>name</th>\n",
              "      <th>display_name</th>\n",
              "      <th>description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>refsnp_id</td>\n",
              "      <td>Variant name</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>refsnp_source</td>\n",
              "      <td>Variant source</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>refsnp_source_description</td>\n",
              "      <td>Variant source description</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr_name</td>\n",
              "      <td>Chromosome/scaffold name</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chrom_start</td>\n",
              "      <td>Chromosome/scaffold position start (bp)</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>chrom_end</td>\n",
              "      <td>Chromosome/scaffold position end (bp)</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>chrom_strand</td>\n",
              "      <td>Strand</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>allele</td>\n",
              "      <td>Variant alleles</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>mapweight</td>\n",
              "      <td>Mapweight</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>validated</td>\n",
              "      <td>Variant supporting evidence</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>allele_1</td>\n",
              "      <td>Ancestral allele</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>minor_allele</td>\n",
              "      <td>Minor allele (ALL)</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>minor_allele_freq</td>\n",
              "      <td>Global minor allele frequency (all individuals)</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>minor_allele_count</td>\n",
              "      <td>Global minor allele count (all individuals)</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>clinical_significance</td>\n",
              "      <td>Clinical significance</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>synonym_name</td>\n",
              "      <td>Synonym name</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>synonym_source</td>\n",
              "      <td>Synonym source</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>synonym_source_description</td>\n",
              "      <td>Synonym source description</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>variation_names</td>\n",
              "      <td>Associated variant names</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>study_name</td>\n",
              "      <td>Study name</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                          name  ... description\n",
              "0                    refsnp_id  ...            \n",
              "1                refsnp_source  ...            \n",
              "2    refsnp_source_description  ...            \n",
              "3                     chr_name  ...            \n",
              "4                  chrom_start  ...            \n",
              "5                    chrom_end  ...            \n",
              "6                 chrom_strand  ...            \n",
              "7                       allele  ...            \n",
              "8                    mapweight  ...            \n",
              "9                    validated  ...            \n",
              "10                    allele_1  ...            \n",
              "11                minor_allele  ...            \n",
              "12           minor_allele_freq  ...            \n",
              "13          minor_allele_count  ...            \n",
              "14       clinical_significance  ...            \n",
              "15                synonym_name  ...            \n",
              "16              synonym_source  ...            \n",
              "17  synonym_source_description  ...            \n",
              "18             variation_names  ...            \n",
              "19                  study_name  ...            \n",
              "\n",
              "[20 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 31
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "SYae9U6cKc4J",
        "outputId": "2587a060-f1cb-4c63-ea52-814eb22bd4e2"
      },
      "source": [
        "dataset.list_filters()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>name</th>\n",
              "      <th>type</th>\n",
              "      <th>description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>link_so_mini_closure</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>link_so_regulation_closure</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>link_so_motif_closure</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr_name</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>start</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>end</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>band_start</td>\n",
              "      <td>drop_down_basic_filter</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>band_end</td>\n",
              "      <td>drop_down_basic_filter</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>marker_start</td>\n",
              "      <td>drop_down_basic_filter</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>marker_end</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>chromosomal_region</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>strand</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>variation_source</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>snp_filter</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>variation_synonym_source</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>snp_synonym_filter</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>study_name</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>study_type</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>phenotype_description</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>phenotype_source</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>pub_short_ref</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>variation_set_name</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>sift_prediction</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>sift_score</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>polyphen_prediction</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>polyphen_score</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>minor_allele_freq</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>minor_allele_freq_second</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>clinical_significance</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>with_validated</td>\n",
              "      <td>boolean</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>with_variation_citation</td>\n",
              "      <td>boolean</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>distance_to_transcript</td>\n",
              "      <td>text</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>ensembl_gene</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>so_consequence_name</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>feature_stable_id</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>regulatory_so_consequence_name</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>motif_name</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>motif_so_consequence_name</td>\n",
              "      <td>list</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                              name                    type description\n",
              "0             link_so_mini_closure                    list            \n",
              "1       link_so_regulation_closure                    list            \n",
              "2            link_so_motif_closure                    list            \n",
              "3                         chr_name                    text            \n",
              "4                            start                    text            \n",
              "5                              end                    text            \n",
              "6                       band_start  drop_down_basic_filter            \n",
              "7                         band_end  drop_down_basic_filter            \n",
              "8                     marker_start  drop_down_basic_filter            \n",
              "9                       marker_end                                    \n",
              "10              chromosomal_region                    text            \n",
              "11                          strand                    text            \n",
              "12                variation_source                    list            \n",
              "13                      snp_filter                    list            \n",
              "14        variation_synonym_source                    list            \n",
              "15              snp_synonym_filter                    list            \n",
              "16                      study_name                    list            \n",
              "17                      study_type                    list            \n",
              "18           phenotype_description                    list            \n",
              "19                phenotype_source                    list            \n",
              "20                   pub_short_ref                    text            \n",
              "21              variation_set_name                    list            \n",
              "22                 sift_prediction                    list            \n",
              "23                      sift_score                    text            \n",
              "24             polyphen_prediction                    list            \n",
              "25                  polyphen_score                    text            \n",
              "26               minor_allele_freq                    text            \n",
              "27        minor_allele_freq_second                    text            \n",
              "28           clinical_significance                    list            \n",
              "29                  with_validated                 boolean            \n",
              "30         with_variation_citation                 boolean            \n",
              "31          distance_to_transcript                    text            \n",
              "32                    ensembl_gene                    list            \n",
              "33             so_consequence_name                    list            \n",
              "34               feature_stable_id                    list            \n",
              "35  regulatory_so_consequence_name                    list            \n",
              "36                      motif_name                    list            \n",
              "37       motif_so_consequence_name                    list            "
            ]
          },
          "metadata": {},
          "execution_count": 30
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DnafXJ0-KrxI"
      },
      "source": [
        "## Exercise\n",
        "\n",
        "Find SNPs in chromosome 5 between position 103000kb and 103001kb"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "76TJ5oRJKCYE"
      },
      "source": [
        "chromosome_number = 5\n",
        "\n",
        "start = 103000000\n",
        "end = 103002000\n",
        "SNPs_of_interest = dataset.query(filters={'chr_name' : str(chromosome_number), 'start': start , 'end' : end},\n",
        "                                  attributes=['refsnp_id',\n",
        "                                              'clinical_significance'])"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 408
        },
        "id": "UrIoTJb7LPgz",
        "outputId": "f6abd716-1b86-4e51-f3dd-92eac1ad9299"
      },
      "source": [
        "SNPs_of_interest"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Variant name</th>\n",
              "      <th>Clinical significance</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>rs975110509</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>rs922327872</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>rs573042954</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>rs992200768</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rs1377715158</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>383</th>\n",
              "      <td>rs758471726</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>384</th>\n",
              "      <td>rs79452628</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>385</th>\n",
              "      <td>rs1428814870</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>386</th>\n",
              "      <td>rs1183028118</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>387</th>\n",
              "      <td>rs1334867951</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>388 rows × 2 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     Variant name  Clinical significance\n",
              "0     rs975110509                    NaN\n",
              "1     rs922327872                    NaN\n",
              "2     rs573042954                    NaN\n",
              "3     rs992200768                    NaN\n",
              "4    rs1377715158                    NaN\n",
              "..            ...                    ...\n",
              "383   rs758471726                    NaN\n",
              "384    rs79452628                    NaN\n",
              "385  rs1428814870                    NaN\n",
              "386  rs1183028118                    NaN\n",
              "387  rs1334867951                    NaN\n",
              "\n",
              "[388 rows x 2 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 38
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "n2OUfSDJLRBt"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}